lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1092,9 +1096,223 @@ class DecomposeAtenMvOp : public OpRewritePattern<AtenMvOp> {
+    prePermuteShape.reserve(nLeadingDims + 5);
+
+    // Shape of 'B' in the comment at the top.
+    SmallVector<Value> postPermuteShape;
+    postPermuteShape.reserve(nLeadingDims + 5);
+
+    SmallVector<Value> outShape;
+    outShape.reserve(nLeadingDims + 3);
+
+    SmallVector<Value> permutation;
+    permutation.reserve(nLeadingDims + 5);
+
+    for (unsigned i = 0; i < nLeadingDims; ++i) {
+      auto dimensionAttr = rewriter.getI64IntegerAttr(i);
+      Value dimensionValue = rewriter.create<ConstantIntOp>(loc, dimensionAttr);
+      Value leadingDimSize =
+          rewriter.createOrFold<AtenSizeIntOp>(loc, inValue, dimensionValue);
+      prePermuteShape.push_back(leadingDimSize);
+      postPermuteShape.push_back(leadingDimSize);
+      outShape.push_back(leadingDimSize);
+      permutation.push_back(dimensionValue);
+
+    }
+
+    const auto inOptionalDType = inType.getOptionalDtype();
+
+    auto getTypeFromShape = [inOptionalDType](auto &&vals) {
+      // Get a vector of integers from a vector of Values.
+      auto getIntShape = [](auto &&vals) {
+        SmallVector<int64_t> shape;
+        shape.reserve(vals.size());
+        for (auto v : vals) {
+          int64_t cst_val;
+          if (matchPattern(v, m_TorchConstantInt(&cst_val))) {
+            shape.push_back(cst_val);
+          } else {
+            shape.push_back(kUnknownSize);
+          }
+        }
+        return shape;
+      };
+
+      const auto intShape = getIntShape(vals);
+      return ValueTensorType::get(vals[0].getContext(),
+                                  llvm::ArrayRef(intShape), inOptionalDType);
+    };
+
+    prePermuteShape.insert(prePermuteShape.end(),
+                           {outC, factor, factor, inH, inW});
+
+    postPermuteShape.insert(postPermuteShape.end(),
+                            {outC, inH, factor, inW, factor});
+
+    outShape.insert(outShape.end(), {outC, outH, outW});
+
+    SmallVector<uint64_t> permutationTail{0, 3, 1, 4, 2};
+    for (uint64_t d : permutationTail) {
+      permutation.push_back(rewriter.create<ConstantIntOp>(
+          loc, rewriter.getI64IntegerAttr(nLeadingDims + d)));
+    }
+
+    auto listType = Torch::ListType::get(Torch::IntType::get(op.getContext()));
+
+
+     Value shapeA =
+         rewriter.create<PrimListConstructOp>(loc, listType, prePermuteShape);
+
+     Value A = rewriter.create<AtenReshapeOp>(
+         loc, getTypeFromShape(prePermuteShape), inValue, shapeA);
+
+     Value permuteDimsOrder = rewriter.create<PrimListConstructOp>(
+         loc, Torch::ListType::get(Torch::IntType::get(op->getContext())),
+         permutation);
+
+     Value B = rewriter.create<AtenPermuteOp>(
+         loc, getTypeFromShape(postPermuteShape), A, permuteDimsOrder);
+
+     Value outShapeList =
+         rewriter.create<PrimListConstructOp>(loc, listType, outShape);
+
+     // TODO(jn) figure out why the type of the returned value must be made
+     // undefined.
```

Comment:
Can you elaborate a bit on what was going wrong here?

Reply:
Below, "finalShape" is  `!torch.vtensor<[3,4,6],si64>`.

But I think "op" has type `!torch.vtensor` (?), and so if I replace 

` rewriter.replaceOpWithNewOp<AtenReshapeOp>(op, finalType, B, outShapeList);`

in the code with 

`rewriter.replaceOpWithNewOp<AtenReshapeOp>(op, finalShape, B, outShapeList);`

I get the error:

```
        python exception: Failure while executing pass pipeline:
        error: unknown: type of return operand 0 ('!torch.vtensor<[3,4,6],si64>') doesn't match function result type ('!torch.vtensor') in function @forward
        note: unknown: see current operation: "func.return"(%11) : (!torch.vtensor<[3,4,6],si64>) -> ()
```

Sorry, my variable names "finalShape" and "finalType" are bad.

Reply:
Ah, so the reason for this is that the rewrite that you do on this operation must replace all original uses of `aten.pixel_shuffle` with values of the same type. Like you said, the op has type `!torch.vtensor`, so we see something like this
```
%0 = torch.aten.pixel_shuffle ... -> !torch.vtensor
%1 = "some_user" %0 : !torch.vtensor -> !torch.int
```
If you don't use the same type, the resulting IR will look like this
```
%0 = torch.aten.pixel_shuffle ... -> !torch.vtensor<[2, 3, 6]xf32>
%1 = "some_user" %0 : !torch.vtensor -> !torch.int
```
and thus there is a type mismatch between %0 and it's user, %1. This is because all operations carry the types of their operands, and you can't change the type of an op operand on the fly.